### DISCUSSION TITLE
Harmonização de dados

### DISCUSSION THREAD

**Discussion by LeonanCarvalho** [BrasilAPI/BrasilAPI#223] [Ideas]
Boa Tarde Pessoal!
Muito bacana a iniciativa!

Tomei conhecimento do projeto ontem através do vídeo do @filipedeschamps  na área da comunidade.

Me identifiquei bastante e gostaria de trazer pra mesa algumas coisas que talvez algum de vocês talvez nunca ouviram falar.

Alguns anos atrás comecei a me envolver devido à necessidades profissionais de uma iniciativa da união europeia chamada [FIWARE](https://www.fiware.org/) , que tinha como rótulo  "Open apais for open minds", bom, hoje é muito mais que isso. e eles tem liderado toda a transformação digital da europa como uma plataforma "open-cloud", aonde tecnologias que estão por vezes travadas dentro dos grandes clouds estariam abertas.

Fiware foi reconhecido também no EUA como a iniciativa mais inovadora e recomendada como adoção para a construção da internet do futuro. 
Eles chamam as tecnologias já consolidadas de "Building Blocks", e pessoal, essa tecnologia é reconhecida e assinada como diretriz pela Camarara de Comécio da Europa através da Connecting Europe Facility (CEF), até o momento 3 building blocks foram definidos, e outros estão em estudo:

- [CEF BDTI](https://www.youtube.com/watch?v=ZaDhzrTb_Nk&feature=youtu.be), apoiar as administrações públicas em sua busca por transformar dados em percepções acionáveis
- [CEF Context Broker](https://www.youtube.com/watch?v=1Cqcrnj9id4), centraliza e fornece contexto para dados provenientes de uma ampla variedade de fontes
- [CEF eArchiving](https://youtu.be/VbR-zvr45so) permitindo a preservação segura, descrição e transmissão de dados digitais

E dentro desses building blocks, o Context Broker é o que mais tem haver com esse projeto,  Context Broker é um Broker que trabalha com uma API de dados de contexto, dados de contexto.

Dados de contexto são dados disponibilizados para seus modelos de visualização(data models) que são renderizados em uma linguagem de modelação como JSON, XML, YML e outros. Os dados de contexto podem consistir em tipos de dados simples, como strings, booleanos, números, arrays e objetos.

Basicamente, tudo em uma API é dado de contexto, mas e se cada pessoa começar a criar suas próprias APIS? no caso do CEP, um pode usar o atributo CEP, outro vai chamar logradouro de rua, e por ai vai, tornaria a vida de todos os DEVs muito mais complicada, teriamos que pra cada nova api ou endpoint reaprender a estrutura de dados.

Foi por isso que foi proposto um modelo de dados chamado NGSI, ou Next Generation Service Interface, desenvolvido inicialmente pela [Open Mobile Aliance](http://www.openmobilealliance.org/), incialmente para dados de sensores, gps, entre outros agora é a estrutura base desse building block que implementam o que chamam de FIWARE NGSI , que é uma especificação da estrutura de dados, e vcs podem ver mais ela por aqui: http://fiware.github.io/specifications/ngsiv2/stable/
Tudo indica que ela se tornará um padrão para a "Europa como uma API", o que esbarra bastante no que vocês tem feito aqui, pelo menos foi o que entendi.

Chega do papo de envagelista e vamos pra prática.

O projeto está no seu estágio inicial , surgiu por uma necessidade básica que é o CEP porém eu vejo que cabe à muitas outras aplicações isso como:

-  FIPE #90  , 
- Dados Meterológicos #88 ,#87
- Informações relacionada à bancos e setor econômico #70  #84  #82, também vejo como uma api para consumo de cotação.
- Dados geográficos como #99  #67  #58

Como vai ficar a apresentação desses dados? Qual guia de estilos será seguido? 
Como profissional, eu não gosto de ficar reinventando a roda, mas utilizando o que foi inventado e adaptando ao meu uso, dando novas aplicações.

A iniciativa que eu falei ali em cima tem trabalhado duro pra isso, e vem criado modelos de dados e publicando eles aqui  https://smartdatamodels.org/ , esse trabalho iniciou assim, aqui no github https://github.com/FIWARE/data-models  e agora se expandiu para uma organização idependente https://github.com/smart-data-models fornecendo modelos de dados padronizados e harmonizados para diversos domínios como Agricultura, Clima, Energia, Meio Ambiente.

Eu acho legal a ideia de termos como base modelos já previamente propostos, e quando ele não se encaixar propor nossas próprias modificações e extenções. Existe ainda além do smart datamodels um projeto mais antigo chamado  https://schema.org/ que fornece referencia para nomes de atributos e estrutura de dados para mais diversos objetos.

O uso de modelos de dados harmonizados é uma necessidade para atender a demanda de interoperabilidade semântica e sintática entre aplicações e dando ao desenvolvedor ferramentas para um rápido desenvolvimento e compatibilidade entre sistemas, e acho que isso é muito relevante para esse proejto. 

Forte Abraço

**Comment by enieber** [comment]
pensando que talvez seja necessário baixar o pdf e salvar em algum lugar para obter alguns desses dados, seria legal salvar os dados em um [dataset aberto como o Brasil.IO](https://brasil.io/datasets/) e usar fazer uma integração com o brasil io como foi sugerido na issue #47

**Comment by isaiasvallejos** [comment]
Sinto que os pontos de padronização são realmente muito importantes! Bora fortalecer isso aqui! 👍 

Provavelmente trabalhar em um tutorial de modelagem/estruturação de requisições, respostas, erros, seria bem legal, porque assim que começarmos a ter mais _endpoints_ (feito por diversos colaboradores) é importantíssimo termos um padrão estabelecido (principalmente para os _reviewers_).

Destaco aqui, só de exemplo:
- **Nome de atributos/propriedades das interfaces**: camelCase, snake_case?
- **Idioma dos atributos**: português ou inglês? 
- **Objeto de erros**: vai ter uma mensagem? Um código de erro? Como vai ser esse código?
- **Padronização de headers**?
- Entre outros...

Eu senti uma complexidade grande nos links que você mandou, será que você poderia dar uma amostra de como podermos aplicar aqui no BrasilAPI? Ficaria legal pra todo mundo entender a ideia.

Uma coisa que é importante de citar é termos que pensar sempre em algo simples e prático para que o repositório fique bastante entendível e "contribuível" para qualquer nível de senioridade, pois esse é o objetivo! 🥇
